File: android/src/main/java/com/gachlab/geolocation/ports/LocationPublisher.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 gachlab

package com.gachlab.geolocation.ports

import com.gachlab.geolocation.BGLocation

/**
 * Port (Roadmap Fase 3, hub → ports): how the hub hands accepted fixes to the sync
 * chiplet (the offline queue + background uploader). The hub depends on this abstraction
 * instead of the concrete `network/PostLocationTask`, so the sync block is swappable and
 * the hub is testable with a fake — and the chiplet (the future `capacitor-event-sink`)
 * is defined by a contract, not a class.
 *
 * Implemented by `network.PostLocationTask` (the sync adapter).
 */
internal interface LocationPublisher {
    /** Enqueue [location] for persistence + (batched) upload. */
    fun add(location: BGLocation)

    /** Release resources (e.g. on reconfigure / service stop). */
    fun shutdown()
}

File: ios/Sources/BackgroundGeolocationCore/Ports/LocationPublisher.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 gachlab

import Foundation

/// Port (Roadmap Fase 3, hub → ports), the iOS twin of the Android
/// `ports/LocationPublisher`: how the hub drives the sync chiplet (the offline queue +
/// background uploader). `BGFacade` depends on this abstraction instead of the concrete
/// `Sync/PostLocationTask` singleton, so the sync block is swappable and the hub is
/// testable with a fake — the chiplet (future `capacitor-event-sink`) is a contract.
///
/// Implemented by `PostLocationTask`.
public protocol LocationPublisher: AnyObject {
    var delegate: PostLocationTaskDelegate? { get set }
    var attachBatterySnapshot: ((BGLocation) -> Void)? { get set }
    var config: BGConfig { get set }
    func start()
    func stop()
    func sync()
    func add(_ location: BGLocation)
}
